src/capymoa/anomaly/online_isolation_forest.py

```diff
@@ -0,0 +1,341 @@
+        else:
+            raise ValueError('Bad type {}'.format(type))
+
+    def learn(self, data: ndarray) -> 'OnlineIsolationTree':
+        # Subsample data in order to improve diversity among trees
+        data: ndarray = data[self.random_generator.random(data.shape[0]) < self.subsample]
+        if data.shape[0] >= 1:
+            # Update the counter of data seen so far
+            self.data_size += data.shape[0]
+            # Adjust depth limit according to data seen so far and branching factor
+            self.depth_limit: float = OnlineIsolationTree.get_random_path_length(self.branching_factor, self.max_leaf_samples,
+                                                                                 self.data_size)
+            # Recursively update the tree
+            self.next_node_index, self.root = self.recursive_build(data) if self.root is None else self.recursive_learn(self.root, data, self.next_node_index)
+        return self
+
+    def recursive_learn(self, node: OnlineIsolationNode, data: ndarray, node_index: int) -> (int, OnlineIsolationNode):
+        # Update the number of data seen so far by the current node
+        node.data_size += data.shape[0]
+        # Update the vectors of minimum and maximum values seen so far by the current node
+        node.min_values: ndarray = vstack([data, node.min_values]).min(axis=0)
+        node.max_values: ndarray = vstack([data, node.max_values]).max(axis=0)
+        # If the current node is a leaf, try to split it
+        if node.children is None:
+            # If there are enough samples to be split according to the max leaf samples and the depth limit has not been
+            # reached yet, split the node
+            if node.data_size >= self.max_leaf_samples*OnlineIsolationTree.get_multiplier(self.type, node.depth) and node.depth < self.depth_limit:
+                # Sample data_size points uniformly at random within the bounding box defined by the vectors of minimum
+                # and maximum values of data seen so far by the current node
+                data_sampled: ndarray = self.random_generator.uniform(node.min_values, node.max_values, size=(node.data_size, data.shape[1]))
+                return self.recursive_build(data_sampled, depth=node.depth, node_index=node_index)
+            else:
+                return node_index, node
+        # If the current node is not a leaf, recursively update all its children
+        else:
+            # Partition data
+            partition_indices: list[ndarray[int]] = self.split_data(data, node.projection_vector, node.split_values)
+            # Recursively update children
+            for i, indices in enumerate(partition_indices):
+                node_index, node.children[i] = self.recursive_learn(node.children[i], data[indices], node_index)
+            return node_index, node
+
+    def recursive_build(self, data: ndarray, depth: int = 0, node_index: int = 0) -> (int, OnlineIsolationNode):
```

Comment:
My IDE complains that ``def foo() -> (int, str)`` should be  ``def foo() -> Tuple[int, str] :``. `Tuple` is from `typing`